code/modules/cm_marines/overwatch.dm

```diff
@@ -1065,6 +489,384 @@
+	data["almayer_cannon_disabled"] = almayer_orbital_cannon.is_disabled ? TRUE : FALSE
+
+	if(current_squad)
+		data["supply_cooldown"] = COOLDOWN_TIMELEFT(current_squad, next_supplydrop)
+		data["marine_list"] = get_marine_data(current_squad)
+		data["squad_data"] = get_squad_data(current_squad)
+
+		var/obj/structure/closet/crate/drop_closet = locate() in current_squad.drop_pad.loc
+		if(drop_closet)
+			data["supply_ready"] = TRUE
+
+	return data
+
+/obj/structure/machinery/computer/overwatch/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(.)
+		return
+
+	var/mob/user = usr
+	if(get_dist(user, src) > 1)
+		return FALSE
+
+	switch(action)
+
+		if("mapview")
+			if(current_mapviewer)
+				update_mapview(1)
+				return TRUE
+			current_mapviewer = usr
+			update_mapview()
+			return TRUE
+
+		if("change_operator")
+			if(operator != usr)
+				if(operator && ishighersilicon(operator))
+					visible_message("[icon2html(src, viewers(src))] [SPAN_BOLDNOTICE("AI override in progress. Access denied.")]")
+					return
+				if(!current_squad || current_squad.assume_overwatch(usr))
+					operator = usr
+				if(ishighersilicon(usr))
+					to_chat(usr, "[icon2html(src, usr)] [SPAN_BOLDNOTICE("Overwatch system AI override protocol successful.")]")
+					current_squad?.send_squad_message("Attention. [operator.name] has engaged overwatch system control override.", displayed_icon = src)
+				else
+					var/mob/living/carbon/human/H = operator
+					var/obj/item/card/id/ID = H.get_idcard()
+					visible_message("[icon2html(src, viewers(src))] [SPAN_BOLDNOTICE("Basic overwatch systems initialized. Welcome, [ID ? "[ID.rank] ":""][operator.name]. Please select a squad.")]")
+					current_squad?.send_squad_message("Attention. Your Overwatch officer is now [ID ? "[ID.rank] ":""][operator.name].", displayed_icon = src)
+			return TRUE
+
+		if("logout")
+			if(current_squad.release_overwatch())
+				if(ishighersilicon(usr))
+					current_squad.send_squad_message("Attention. [operator.name] has released overwatch system control. Overwatch functions deactivated.", displayed_icon = src)
+					to_chat(usr, "[icon2html(src, usr)] [SPAN_BOLDNOTICE("Overwatch system control override disengaged.")]")
+				else
+					var/mob/living/carbon/human/H = operator
+					var/obj/item/card/id/ID = H.get_idcard()
+					current_squad.send_squad_message("Attention. [ID ? "[ID.rank] ":""][operator ? "[operator.name]":"sysadmin"] is no longer your Overwatch officer. Overwatch functions deactivated.", displayed_icon = src)
+					visible_message("[icon2html(src, viewers(src))] [SPAN_BOLDNOTICE("Overwatch systems deactivated. Goodbye, [ID ? "[ID.rank] ":""][operator ? "[operator.name]":"sysadmin"].")]")
+			operator = null
+			current_squad = null
+			if(cam && !ishighersilicon(usr))
+				usr.reset_view(null)
+			cam = null
+			return TRUE
+
+		if("pick_squad")
+			if(operator != usr)
+				return FALSE
```

Comment:
do these checks in the UI, if they're already done in the UI you do not need them here